[Docs] : how to create expectations while exploring data in a notebook without using a data context (#1514)
great-expectations/great_expectations: docs/how_to_guides/creating_and_editing_expectations/how_to_create_expectations_while_exploring_data_in_a_notebook_without_using_a_data_context.rst

How to create Expectations while exploring data in a notebook, without using a Data Context.
============================================================================================

Building :ref:`Expectations` as you conduct exploratory data analysis is a great way to ensure that your insights about data processes and pipelines remain part of your team's knowledge.

This guide will help you get started quickly in Great Expectations, without even setting up a :ref:`Data Context`. All you need is a notebook and some data.

.. admonition:: Prerequisites: This how-to guide assumes you have already:

    - Installed Great Expectations (e.g. ``pip install great_expectations``)
    - Have access to a notebook (e.g. ``jupyter notebook``, ``jupyter lab``, etc.)

>     - Be able to access data from your notebook

Reviewer: Parallelism
```suggestion
    - Obtained a sample of data to use for exploration
```

>     - Nothing else. Unlike most how-to guides, these instructions do *not* assume that you have configured a Data Context by running ``great_expectations init``.

Reviewer: ```suggestion
```

Author: I very much want to keep this item.

Once we have three or four of them, I'm planning to link them from (and to) a "magic free" section in the Quick Start tutorial that will allow users to get a taste of Great Expectations without even having to set up a Data Context.

Of course, a big part of that tutorial will be selling users on the value of Data Contexts, too.

Steps
-----

All of these steps take place within your notebook:

1. **Import Great Expectations.**

    .. code-block:: python

        import great_expectations as ge

2. **Load some data.**

    The simplest way to do this is with ``read_csv``.

    .. code-block:: python

        my_df = ge.read_csv("my_data_directory/titanic.csv")

    This method behaves exactly the same as ``pandas.read_csv``, so you can add parameters to parse your file:
    
    .. code-block:: python

        my_df = ge.read_csv(
            "my_data_directory/my_messy_data.csv",
            sep="\t",
            skiprows=3
        )

>     Similarly wrapped versions of other pandas methods (``read_excel``, ``read_table``, ``read_parquet``, ``read_pickle``, ``read_json``, etc.) are also available. Please see the <great_expectations utils module> for details.

Reviewer: broken link

Author: We should talk about linking to the module docs. As they stand, I don't know what to trust there.

>     If you wish to load data from somewhere else (e.g. from a SQL database or blob store), please fetch a copy of the data locally, or :ref:`configure a Datasource <tutorials__getting_started__connect_to_data>`. (In the future, you will likely want to configure Datasources anyway, to take advantage of more of Great Expectations' advanced features.)

Reviewer: I don't think the redirect should be to configure a datasource, but instead to use the raw classes (`SparkDFDataset` or `SqlAlchemyDataset`)

Author: Agreed. Again, I'm avoiding module docs for the moment, because they don't seem to be consistently implemented today. If there's a way to whip them into shape quickly, great. Otherwise, it's a pile of work I'm reluctant to take on right now.

    As another option, if you have already instantiated a ``pandas.Dataframe``, you can use ``from_pandas``:

    .. code-block:: python

        my_df = ge.from_pandas(
            my_pandas_dataframe
        )

>     This method will convert your boring old pandas ``DataFrame`` into a new and exciting great_expectations ``PandasDataAsset``. The two classes are absolutely identical, except that ``PandasDataAsset`` has access to Great Expectations' methods.

Reviewer: ```suggestion
    This method will convert your boring old pandas ``DataFrame`` into a new and exciting great_expectations ``PandasDataset``. The two classes are absolutely identical, except that ``PandasDataset`` has access to Great Expectations' methods.
```

3. **Explore your data and add Expectations.**

>     Each of the methods in step 1 will produce ``my_df``, a ``PandasDataAsset``. ``PandasDataAsset`` is a subclass of ``pandas.DataFrame``, which means that you can use all of pandas' normal methods on it.

Reviewer: ```suggestion
    Each of the methods in step 1 will produce ``my_df``, a ``PandasDataset``. ``PandasDataset`` is a subclass of ``pandas.DataFrame``, which means that you can use all of pandas' normal methods on it.
```

    .. code-block:: python

        my_df.head()
        my_df.Sex.value_counts()
        my_df[my_df.Sex=="male"].head()
        # etc., etc. 
        
    In addition, ``my_df`` has access to a wide array of Expectations. You can see the full list :ref:`here <Glossary of Expectations>`. By convention, every Expectation method name starts with the name ``expect_...``, so you can quickly access the full list with tab-based autocomplete:

>     .. image:: ../../images/expectation_autocomplete.gif

Reviewer: <3 <3 <3

Author: +1

    |

>     When you invoke an Expectation, it will immediately be validated against your data. The returned object will contain the result and a list of exceptions. This instant feedback helps you zero in on exceptions very quickly, taking a lot of the guesswork out of data exploration.

Reviewer: ```suggestion
    When you invoke an Expectation, it will immediately be validated against your data. The returned object will contain the result and a list of unexpected values. This instant feedback helps you zero in on unexpected data very quickly, taking a lot of the guesswork out of data exploration.
```

Author: ooh. Good catch.

>     .. image:: ../../images/expectation_notebook_interactive_loop.gif

Reviewer: <3 <3 <3

Author: +1

        Note how ``success`` switches from ``false`` to ``true`` once ``mostly=.99`` is added.

    |

4. **Review your Expectations.**

    As you run Expectations in your notebook, ``my_df`` will build up a running list of Expectations. By default, Great Expectations will recognize and replace duplicate Expectations, so that only the most recent version is stored. (See :ref:`Determining duplicate results` below for details.)

    You can get the config file for your Expectations by running:

    .. code-block:: python
    
        my_df.get_expectation_suite()

    which will return an :ref:`ExpectationSuite` object.

    By default, ``get_expectation_suite()`` only returns Expectations with ``success=True`` on their most recent validation. You can override this behavior with:
    
    .. code-block:: python

        my_df.get_expectation_suite(discard_failed_expectations=False)

5. **Save your Expectation Suite.**

    Expectation Suites can be serialized as JSON objects, so you can save your Expectation Suite like this:

    .. code-block:: python
    
        import json

        with open( "my_expectation_file.json", "w") as my_file:
            my_file.write(

>                 json.dumps(my_df.get_expectation_suite().to_json_dict())

Reviewer: Do you like using `to_json_dict` here? Other options might be to use the schema directly or to `print()` the object

Author: hm.

* `to_json_dict` didn't give me any concerns. I like that it's self-evident what the method is doing. I agree this is a tad clunky, but that's partly deliberate (See #1544).
* The notion of using `print()` feels off to me. I think that python users think of `print` as their "write to console" method. I'm pretty sure that using it to write elsewhere will confuse people.
* I'm not sure what you mean by "using the schema directly." Can you clarify?

Reviewer: Resolved synchronously.
